pss/outbox/outbox.go

```diff
@@ -0,0 +1,130 @@
+package outbox
+
+import (
+	"errors"
+
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/pss/message"
+)
+
+type Config struct {
+	NumberSlots int
+	NumWorkers  int
+	Forward     forwardFunction
+}
+
+type Outbox struct {
+	forwardFunc forwardFunction
+	queue       []*outboxMsg
+	slots       chan int
+	process     chan int
+	workers     chan int
+	stopC       chan struct{}
+}
+
+type forwardFunction func(msg *message.Message) error
+
+var ErrOutboxFull = errors.New("outbox full")
+
+const defaultOutboxWorkers = 100
+
+func NewOutbox(config *Config) *Outbox {
+	if config.NumWorkers == 0 {
+		config.NumWorkers = defaultOutboxWorkers
+	}
+	outbox := &Outbox{
+		forwardFunc: config.Forward,
+		queue:       make([]*outboxMsg, config.NumberSlots),
+		slots:       make(chan int, config.NumberSlots),
+		process:     make(chan int),
+		workers:     make(chan int, config.NumWorkers),
```

Comment:
Workers is just a semaphore, no need to have int type, it can be struct{}. The value passed to this channel is never used. I would consider using a different name then workers as it has only the limiting effect.

Also, this channel is only used in processOutbox. Consider creating it there instead to be a field in Outbox struct.